scripts/histmakers/mw_with_mu_eta_pt_VETOEFFI.py

```diff
@@ -0,0 +1,160 @@
+
+era = args.era
+datasets = getDatasets(maxFiles=args.maxFiles,
+                       filt=args.filterProcs,
+                       excl=args.excludeProcs, 
+                       nanoVersion="v9", base_path=args.dataPath, oneMCfileEveryN=args.oneMCfileEveryN,
+                       extended = "msht20an3lo" not in args.pdfs,
+                       era=era)
+
+# custom template binning
+template_neta = int(args.eta[0])
+template_mineta = args.eta[1]
+template_maxeta = args.eta[2]
+logger.info(f"Eta binning: {template_neta} bins from {template_mineta} to {template_maxeta}")
+template_npt = int(args.pt[0])
+template_minpt = args.pt[1]
+template_maxpt = args.pt[2]
+logger.info(f"Pt binning: {template_npt} bins from {template_minpt} to {template_maxpt}")
+
+# standard regular axes
+axis_eta = hist.axis.Regular(template_neta, template_mineta, template_maxeta, name = "eta", overflow=False, underflow=False)
+axis_pt = hist.axis.Regular(template_npt, template_minpt, template_maxpt, name = "pt", overflow=False, underflow=False)
+axis_charge = common.axis_charge
+axis_passVeto = hist.axis.Boolean(name = "passVeto")
+
+nominal_axes = [axis_eta, axis_pt, axis_charge, axis_passVeto]
+nominal_cols = ["postFSRmuon_eta0", "postFSRmuon_pt0", "postFSRmuon_charge0", "passVeto"]
+
+# sum those groups up in post processing
+groups_to_aggregate = args.aggregateGroups
+
+#qcdScaleByHelicity_helper = wremnants.theory_corrections.make_qcd_uncertainty_helper_by_helicity()
+
+logger.info("Running with no scale factors")
+
+pileup_helper = wremnants.make_pileup_helper(era = era)
+vertex_helper = wremnants.make_vertex_helper(era = era)
+
+calib_filepaths = common.calib_filepaths
+closure_filepaths = common.closure_filepaths
+
+diff_weights_helper = ROOT.wrem.SplinesDifferentialWeightsHelper(calib_filepaths['tflite_file']) if (args.muonScaleVariation == 'smearingWeightsSplines' or args.validationHists) else None
+
+mc_jpsi_crctn_helper, data_jpsi_crctn_helper, jpsi_crctn_MC_unc_helper, jpsi_crctn_data_unc_helper = muon_calibration.make_jpsi_crctn_helpers(args, calib_filepaths, make_uncertainty_helper=True)
+
+z_non_closure_parametrized_helper, z_non_closure_binned_helper = muon_calibration.make_Z_non_closure_helpers(args, calib_filepaths, closure_filepaths)
+
+mc_calibration_helper, data_calibration_helper, calibration_uncertainty_helper = muon_calibration.make_muon_calibration_helpers(args)
+
+smearing_helper, smearing_uncertainty_helper = (None, None) if args.noSmearing else muon_calibration.make_muon_smearing_helpers()
+
+bias_helper = muon_calibration.make_muon_bias_helpers(args) if args.biasCalibration else None
+
+procsWithTheoryCorr = [d.name for d in datasets if d.name in common.vprocs]
+if len(procsWithTheoryCorr):
+    corr_helpers = theory_corrections.load_corr_helpers(procsWithTheoryCorr, args.theoryCorr, allowMissingTheoryCorr=args.allowMissingTheoryCorr)
+else:
+    corr_helpers = {}
+
+smearing_weights_procs = []
+
+def build_graph(df, dataset):
```

Comment:
In the long run this would be more maintainable if some of the common code between this and the mw script is moved into helper functions

Reply:
I know, the fact is that sometimes you need a custom loop, and having common pieces of code is not always viable. 
In fact, this histmaker already uses common functions to apply cuts, but you can't easily have a common function which gives you a customisable flow with cuts, defines, and helpers. For that purpose I think the correct solution is just writing an independent loop where you put what you need, even at the cost of some partial duplication. In fact I was not even fully sure I should commit this histmaker, since those MC efficiencies are not currently used directly in the analysis. Perhaps it should be moved in some test folder for utility plots.

Reply:
I know. But there is a lot of code that is copy paste, that could be grouped together.

Reply:
Probably, and if this is possible I would be in favour.
However, quite often I noticed that hiding too much code into a common set of pieces goes against one's needs, since it either forces you to further generalise those pieces, such that the original purpose of having a general function or sets of function gets broken, or in any case sometimes requires you to fall back to rewriting the loop anyway since what you need might not be possible to achieve with the general code. Even if in this particular case it might be doable, tomorrow I might need something totally different, so I just don't know how much we have to gain at the moment by generalising.
It is essentially the same reason we kept thew Wlike and mW histomakers separate, they are logically connected but there are too many practical differences for which a common script is not suited for.
CMGTools was a largely general code, where you set the cuts, histograms, and definitions outside of the loop backend, which would work for any analysis, but eventually this power was also its weakness when you needed a higher level of customization or special features (sorry, despite how much I hated CMGTools before getting used to it I couldn't help but making this statement :) )